### ISSUE TITLE
Flag to decide whether to expose externally or not

### ISSUE THREAD

**Issue by surajssd** [surajssd/kapp#22]
So right now there is no defining way to know which service to expose outside cluster using `ingress`.

As the implementation goes, it is checked if the service type is `LoadBalancer` only then it is exposed and yesterday during demo I realized service of any type can be exposed using `route` or `ingress`. Also as yesterday @jstrachan suggested adding a flag to the service spec which decides whether to expose it or not can be used. So proposing on adding a flag called `expose` to the root level `services` section.

Following is the snippet of how the `service` spec will look like:

```yaml
services:
- name: wordpress
  expose: true
  ports:
  - port: 8080
    targetPort: 80
  rules:
  - host: minikube.local
```

If user specifies this much a service is created with the ports given, also the flag `expose` is set we will create `ingress`. Since this is for Kubernetes we will need a rule specifying host. (**Note**: If this was for OpenShift we would not need that. Let's put the discussion for OpenShift routes for later.)

Now if the user has specified only one rule and host in it, then we expand the rest of `ingress` spec for user, which essentially becomes

```yaml
services:
- name: wordpress
  expose: true
  ports:
  - port: 8080
    targetPort: 80
  rules:
  - host: minikube.local
    http:
      paths:
      - backend:
          serviceName: wordpress
          servicePort: 8080
        path: /
```

Now if user wants to specify her rule she can decide to do it writing everything. I think this makes for intelligent default.

**Comment by kadel** [comment]
This also means that SericeSpec and IngressSpec will be merged together?

**Comment by jstrachan** [comment]
Incidentally there's Ingress, Route, NodePort or LoadBalancer as the main 4 strategies to expose services externally. 

In the fabric8 ecosystem we tend to use this little microservice to take care of that for us: https://github.com/fabric8io/exposecontroller/ it auto-detects openshift and defaults to Route there or can be configured in a ConfigMap to pick LoadBalancer (if on public cloud), NodePort if running minikube/minishift or defaults to Ingress on kubernetes.

**Comment by kadel** [comment]
it might be better to do Ingress as a list, so you can create multiple ingress for one service

```
- name: wordpress
  expose: true
  ports:
  - port: 8080
    targetPort: 80
    ingress:
     - name: nontls
       rules:
         - host: minikube.local
     - name: tls
       tls: asdf
       rules:
         - host: minikube.local
```

**Comment by containscafeine** [comment]
picking this up

**Comment by containscafeine** [comment]
@kadel -

Did you mean -

```yaml
- name: wordpress
  expose: true
  ports:
  - port: 8080
    targetPort: 80
  ingress:
  - name: nontls
    rules:
    - host: minikube.local
  - name: tls
    tls: asdf
    rules:
    - host: minikube.local
```

**Comment by containscafeine** [comment]
Okay, so after discussing this with @surajssd yesterday, we think we need one more round of discussion around this.

- We had previously decided that `expose: true` will not work with Kubernetes, but it can; we found out that `host` is an optional field in Kubernetes after all, and the following is a valid Ingress resourse

```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: no-rules
spec:
  backend:
    serviceName: s1
    servicePort: 80
```
Now, we need to discuss behavior around this. Fr this Ingree resource, all the traffic on all paths for a given IP by the Ingress Controller will be routed to the specified default backend.
But in OpenShift, the cluster will create an auto generated URL.

- Where do we merge the `IngressSpec`? Do we embed a `Ingress []IngressSpec` in the `ServiceSpec` as mentioned in the comment above (which is what @surajssd and me thought after day before's discussion), or should we embed this with `ServicePort` spec, like @kadel's comment above. Going the latter way, we can specify an ingress resource per service port, and going with former, we will have to fail if multiple ports in one service is provided. We need to think more about this :(
